core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushdownFilterIntoWindow.java

```diff
@@ -0,0 +1,185 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.Session;
+import io.trino.matching.Capture;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.metadata.FunctionId;
+import io.trino.metadata.Metadata;
+import io.trino.spi.predicate.Domain;
+import io.trino.spi.predicate.Range;
+import io.trino.spi.predicate.TupleDomain;
+import io.trino.spi.predicate.ValueSet;
+import io.trino.spi.type.TypeOperators;
+import io.trino.sql.ExpressionUtils;
+import io.trino.sql.planner.DomainTranslator;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.TypeProvider;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.FilterNode;
+import io.trino.sql.planner.plan.TopNRowNumberNode;
+import io.trino.sql.planner.plan.ValuesNode;
+import io.trino.sql.planner.plan.WindowNode;
+import io.trino.sql.tree.BooleanLiteral;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.QualifiedName;
+
+import java.util.Optional;
+import java.util.OptionalInt;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.SystemSessionProperties.isOptimizeTopNRowNumber;
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.spi.predicate.Marker.Bound.BELOW;
+import static io.trino.spi.predicate.Range.range;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.sql.planner.DomainTranslator.fromPredicate;
+import static io.trino.sql.planner.plan.Patterns.filter;
+import static io.trino.sql.planner.plan.Patterns.source;
+import static io.trino.sql.planner.plan.Patterns.window;
+import static java.lang.Math.toIntExact;
+
+public class PushdownFilterIntoWindow
+        implements Rule<FilterNode>
+{
+    private final Capture<WindowNode> childCapture;
+    private final Pattern<FilterNode> pattern;
+
+    private final Metadata metadata;
+    private final DomainTranslator domainTranslator;
+    private final FunctionId rowNumberFunctionId;
+    private final TypeOperators typeOperators;
+
+    public PushdownFilterIntoWindow(Metadata metadata, TypeOperators typeOperators)
+    {
+        this.metadata = metadata;
+        this.domainTranslator = new DomainTranslator(metadata);
+        this.rowNumberFunctionId = metadata.resolveFunction(QualifiedName.of("row_number"), ImmutableList.of()).getFunctionId();
+        this.childCapture = newCapture();
+        this.pattern = filter()
+                .with(source().matching(window()
+                        .matching(window -> window.getWindowFunctions().size() == 1 && getOnlyElement(window.getWindowFunctions().values()).getResolvedFunction().getFunctionId().equals(rowNumberFunctionId))
```

Comment:
I think there's one comment not yet applied: the pattern should contain the following condition for the WindowNode: `.matching(window -> window.getOrderingScheme().isPresent())`.

Reply:
Sorry. I missed the reply to this [comment](https://github.com/trinodb/trino/pull/4242/files#r527894916). 

Therefore we can remove the following code from `PushdownFilterIntoWindow` because that transformation is handled by `PushdownFilterIntoRowNumber` after `ReplaceWindowWithRowNumber`. That allows us to add the pattern `.matching(window -> window.getOrderingScheme().isPresent()` in `PushdownFilterIntoWindow`. 

```
if (upperBound.getAsInt() <= 0) {
  return Result.ofPlanNode(new ValuesNode(node.getId(), node.getOutputSymbols(), ImmutableList.of()));
}
```

Reply:
Sorry. I missed the reply to this [comment](https://github.com/trinodb/trino/pull/4242/files#r527894916). 

Therefore we can remove the following code from `PushdownFilterIntoWindow` because that transformation is handled by `PushdownFilterIntoRowNumber` after `ReplaceWindowWithRowNumber`. That allows us to add the pattern `.matching(window -> window.getOrderingScheme().isPresent()` in `PushdownFilterIntoWindow`. 

```
if (upperBound.getAsInt() <= 0) {
  return Result.ofPlanNode(new ValuesNode(node.getId(), node.getOutputSymbols(), ImmutableList.of()));
}
```

Reply:
Hmm, after this change the test `TestWindowFilterPushDown` starts failing. Can we ensure to run `ReplaceWindowWithRowNumber` and `PushdownFilterIntoRowNumber` all time?

https://github.com/trinodb/trino/pull/6507/checks?check_run_id=1653865946

Reply:
Here's what happened. Recently,  a change has been merged in which `TopNRowNumberNode` was extended to handle other ranking operations with limited output. Now there is `TopNRankingNode` with type as an enum.
See: https://github.com/trinodb/trino/commit/ab4d9cd48fde1371a26bad24b1d11180daafe29b

The Optimizer which you are migrating was updated accordingly and now it can handle not only WindowNode with `row_number()` function but also with `rank()` and `dense_rank()`. However, the new rules you added only handle `row_number()`. There's need to update the rules accordingly so that they reflect the full capacity of `WindowFilterPushdown` optimizer. That is, support also `rank()` and `dense_rank()` in WindowNode.

@erichwang could you please confirm that the above explanation is accurate or correct it?

Reply:
Correct, TopNRowNumberNode was generalized to TopNRankingNode (with 3 enums to differentiate). Note that topn dense_rank() has not been implemented yet (and so you will not see any planner rules to enable this yet), but the node scaffolding is set up there to receive it.

Reply:
@Lewuathe, let me know if anything doesn't seem clear, happy to explain more. It should be mostly mechanical change here.

Reply:
Okay. Thanks for the explanation. I'll try to extend `PushdownLimitIntoWindow` rule to be aware of `rank` function as well.

Reply:
Now the rule supports `rank` as well.